MSAL/src/native_auth/cache/MSALNativeAuthCacheAccessor.swift

```diff
@@ -0,0 +1,181 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import Foundation
+@_implementationOnly import MSAL_Private
+
+class MSALNativeAuthCacheAccessor: MSALNativeAuthCacheInterface {
+
+    private let tokenCacheAccessor: MSIDDefaultTokenCacheAccessor = {
+        let dataSource = MSIDKeychainTokenCache()
+        return MSIDDefaultTokenCacheAccessor(dataSource: dataSource, otherCacheAccessors: [])
+    }()
+
+    private let accountMetadataCache: MSIDAccountMetadataCacheAccessor = MSIDAccountMetadataCacheAccessor(dataSource: MSIDKeychainTokenCache())
+    private let externalAccountProvider: MSALExternalAccountHandler = MSALExternalAccountHandler()
+    private let validator = MSIDTokenResponseValidator()
+
+    func getTokens(
+        account: MSALAccount,
+        configuration: MSIDConfiguration,
+        context: MSIDRequestContext) throws -> MSALNativeAuthTokens {
+            let accountConfiguration = try getAccountConfiguration(configuration: configuration, account: account)
+            let idToken = try tokenCacheAccessor.getIDToken(
+                forAccount: account.lookupAccountIdentifier,
+                configuration: accountConfiguration,
+                idTokenType: MSIDCredentialType.MSIDIDTokenType,
+                context: context)
+            let refreshToken = try tokenCacheAccessor.getRefreshToken(
+                withAccount: account.lookupAccountIdentifier,
+                familyId: nil,
+                configuration: accountConfiguration,
+                context: context)
+            let accessToken = try tokenCacheAccessor.getAccessToken(
+                forAccount: account.lookupAccountIdentifier,
+                configuration: accountConfiguration,
+                context: context)
+            return MSALNativeAuthTokens(accessToken: accessToken, refreshToken: refreshToken, rawIdToken: idToken.rawIdToken)
+        }
+
+    func getAccount(
```

Comment:
I noticed that MSID* and MSAL* classes are mixed in this class:
- getAccount method returns MSIDAccount
- getAllAccounts returns [MSALAccount]

Should we separate layers and don't mix them? MSAL classes are just public wrappers around MSID classes, so I think usually for your private logic you should use MSID layer only.

Reply:
Thanks Sergei, we removed the unused "getAccount " method on ADO. We will reflect this change also on GitHub